In [29]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import test
from util import *

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- DP]art I ---</h2></article>


In [ ]:
tests = [
    {
        "name": "Example",
        "s": """
                #####
                .####
                .####
                .####
                .#.#.
                .#...
                .....

                #####
                ##.##
                .#.##
                ...##
                ...#.
                ...#.
                .....

                .....
                #....
                #....
                #...#
                #.#.#
                #.###
                #####

                .....
                .....
                #.#..
                ###..
                ###.#
                ###.#
                #####

                .....
                .....
                .....
                #....
                #.#..
                #.#.#
                #####
        """,
        "expected": 3,
    },
]


def part_I(s: str) -> int:
    inp_iter = (
        [l.strip() for l in b.splitlines()]
        for b in re.split(r"(?:\r?\n){2,}", s.strip())
    )

    locks, keys = [], []

    for block in inp_iter:
        if block[0][0] == "#":
            rotated = ["".join(row) for row in zip(*block[1:-1][::1])]
            locks.append([l.count("#") for l in rotated])
        else:
            rotated = ["".join(row) for row in zip(*block[1:-1][::-1])]
            keys.append([l.count("#") for l in rotated])

    return sum(
        1
        for lock, key in product(locks, keys)
        if all(l + k <= 5 for l, k in zip(lock, key))
    )


@test(tests=tests[:])
def partI_test(s: str) -> int:
    return part_I(s)


Test Example passed, for partI_test.
Success


In [31]:
with open("../input/day25.txt") as f:
    puzzle = f.read()

print(f"Part I: {part_I(puzzle)}")

Part I: 3090


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>3090</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>


<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>You and The Historians crowd into the office, startling the Chief Historian awake! The Historians all take turns looking confused until one asks where he's been for the last few months.</p>
<p>"I've been right here, working on this high-priority request from Santa! I think the only time I even stepped away was about a month ago when I went to grab a cup of coffee..."</p>
<p>Just then, the Chief notices the time. "Oh no! I'm going to be late! I must have fallen asleep trying to put the finishing touches on this <em>chronicle</em> Santa requested, but now I don't have enough time to go visit the last 50 places on my list and complete the chronicle before Santa leaves! He said he needed it before tonight's sleigh launch."</p>
<p>One of The Historians holds up the list they've been using this whole time to keep track of where they've been searching. Next to each place you all visited, they checked off that place with a <em class="star">star</em>. Other Historians hold up their own notes they took on the journey; as The Historians, how could they resist writing everything down while visiting all those historically significant places?</p>
<p>The Chief's eyes get wide. "With all this, we might just have enough time to finish the chronicle! Santa said he wanted it wrapped up with a bow, so I'll call down to the wrapping department and... hey, could <em>you</em> bring it up to Santa? I'll need to be in my seat to watch the sleigh launch by then."</p>
<p>You nod, and The Historians quickly work to collect their notes into the final set of pages for the chronicle.</p>
</article>


<link href="style.css" rel="stylesheet"></link>


<link href="style.css" rel="stylesheet"></link>
<main>
<article>
<p>Using the notes from the places marked by all <span class="day-success">fifty stars</span>, The Historians finish the chronicle, wrap it, and give it to you so you can bring it to Santa before the big sleigh launch.</p><p>Santa is already in the sleigh making the final launch preparations when you arrive. You try to hand him the chronicle, but he doesn't take it. "Ho ho ho," he laughs to himself. "That gift isn't for me - it's for <em>you</em>. That chronicle is a record of all the places you've been and people you've helped over the last decade. Thank you for everything." With that, Santa takes off in his sleigh to deliver the rest of this year's presents.</p>
<p class="aside">Congratulations!  You've finished every puzzle in Advent of Code 2024!  I hope you had as much fun solving them as I had making them for you.  I'd love to hear about your adventure; you can get in touch with me via contact info on <a href="https://was.tl/" target="_blank">my website</a> or through <a href="https://bsky.app/profile/was.tl" target="_blank">Bluesky</a> or <a href="https://hachyderm.io/@ericwastl" target="_blank">Mastodon</a>.</p>
<p class="aside">If you'd like to see more things like this in the future, please consider <a href="/2024/support" target="_blank">supporting</a> Advent of Code and sharing it with others.</p>
<p class="aside">I've <span style="border-bottom:1px dotted #ffff66;" title="Yep, just like that.  There's at least one in the description for each day.">highlighted</span> the easter eggs in each puzzle, just in case you missed any.  Hover your mouse over them, and the easter egg will appear.</p>
<p>You can <span class="share">[Share<span class="share-content">on
  <a href="https://bsky.app/intent/compose?text=I+just+completed+all+25+days+of+Advent+of+Code+2024%21+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F" target="_blank">Bluesky</a>
  <a href="https://twitter.com/intent/tweet?text=I+just+completed+all+25+days+of+Advent+of+Code+2024%21&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I+just+completed+all+25+days+of+Advent+of+Code+2024%21+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this moment with your friends or <a href="/2024">[Go Check on Your Calendar]</a>.
</p></article>
<style>
.calendar-bkg {
  background: linear-gradient(to bottom, rgba(255,255,255,0) 0%,rgba(255,255,255,.3) 100%);
}
.sf {
  position: fixed;
  animation: anim-sf0 linear 20s infinite;
  z-index: -1;
}
.sf:before {
  content: "*";
}
.sf.sf0 { color: #ffffff; }
.sf.sf1 { color: #cccccc; animation-name: anim-sf1; }
.sf.sf2 { color: #999999; animation-name: anim-sf2; }
.sf.sf3 { color: #666666; animation-name: anim-sf3; }
@keyframes anim-sf0 {
  0%   { transform: translate(  0,     0) rotate(0)     scale(1.0, 1.0); }
  95%  { transform: translate(8em, 100vh) rotate(1turn) scale(1.0, 1.0); }
  100% { transform: translate(8em, 100vh) rotate(1turn) scale(0.0, 0.0); }
}
@keyframes anim-sf1 {
  0%   { transform: translate(  0,     0) rotate(0)     scale(1.0, 1.0); }
  95%  { transform: translate(6em, 100vh) rotate(1turn) scale(1.0, 1.0); }
  100% { transform: translate(6em, 100vh) rotate(1turn) scale(0.0, 0.0); }
}
@keyframes anim-sf2 {
  0%   { transform: translate(  0,     0) rotate(0)     scale(1.0, 1.0); }
  95%  { transform: translate(4em, 100vh) rotate(1turn) scale(1.0, 1.0); }
  100% { transform: translate(4em, 100vh) rotate(1turn) scale(0.0, 0.0); }
}
@keyframes anim-sf3 {
  0%   { transform: translate(  0,     0) rotate(0)     scale(1.0, 1.0); }
  95%  { transform: translate(2em, 100vh) rotate(1turn) scale(1.0, 1.0); }
  100% { transform: translate(2em, 100vh) rotate(1turn) scale(0.0, 0.0); }
}
</style>
<div class="sf sf3" style="top:-4.61em; left:86.22vw; animation-delay:-6.09s;"></div><div class="sf sf0" style="top:-1.66em; left:6.96vw; animation-delay:-2.77s;"></div><div class="sf sf1" style="top:-2.41em; left:74.59vw; animation-delay:-7.81s;"></div><div class="sf sf2" style="top:-3.49em; left:36.76vw; animation-delay:-17.10s;"></div><div class="sf sf1" style="top:-2.10em; left:20.26vw; animation-delay:-10.37s;"></div><div class="sf sf3" style="top:-4.08em; left:51.05vw; animation-delay:-13.47s;"></div><div class="sf sf2" style="top:-3.38em; left:93.22vw; animation-delay:-6.17s;"></div><div class="sf sf1" style="top:-2.22em; left:77.69vw; animation-delay:-18.29s;"></div><div class="sf sf1" style="top:-2.08em; left:43.44vw; animation-delay:-1.96s;"></div><div class="sf sf1" style="top:-2.04em; left:38.31vw; animation-delay:-16.65s;"></div><div class="sf sf0" style="top:-1.72em; left:80.66vw; animation-delay:-3.84s;"></div><div class="sf sf0" style="top:-1.11em; left:82.04vw; animation-delay:-8.80s;"></div><div class="sf sf1" style="top:-2.44em; left:19.80vw; animation-delay:-5.88s;"></div><div class="sf sf0" style="top:-1.56em; left:35.55vw; animation-delay:-10.44s;"></div><div class="sf sf2" style="top:-3.57em; left:49.60vw; animation-delay:-3.79s;"></div><div class="sf sf3" style="top:-4.13em; left:90.29vw; animation-delay:-7.50s;"></div><div class="sf sf1" style="top:-2.69em; left:22.01vw; animation-delay:-18.91s;"></div><div class="sf sf2" style="top:-3.85em; left:68.93vw; animation-delay:-12.69s;"></div><div class="sf sf1" style="top:-2.67em; left:16.82vw; animation-delay:-11.71s;"></div><div class="sf sf3" style="top:-4.06em; left:14.98vw; animation-delay:-7.93s;"></div><div class="sf sf2" style="top:-3.17em; left:25.63vw; animation-delay:-9.28s;"></div><div class="sf sf3" style="top:-4.60em; left:5.18vw; animation-delay:-14.60s;"></div><div class="sf sf1" style="top:-2.53em; left:15.74vw; animation-delay:-6.23s;"></div><div class="sf sf1" style="top:-2.93em; left:45.77vw; animation-delay:-11.89s;"></div><div class="sf sf3" style="top:-4.33em; left:78.86vw; animation-delay:-5.23s;"></div><div class="sf sf1" style="top:-2.21em; left:89.24vw; animation-delay:-2.29s;"></div><div class="sf sf0" style="top:-1.06em; left:15.75vw; animation-delay:-19.32s;"></div><div class="sf sf3" style="top:-4.24em; left:64.31vw; animation-delay:-2.95s;"></div><div class="sf sf3" style="top:-4.08em; left:77.07vw; animation-delay:-6.34s;"></div><div class="sf sf3" style="top:-4.71em; left:62.06vw; animation-delay:-1.74s;"></div><div class="sf sf0" style="top:-1.95em; left:76.33vw; animation-delay:-3.14s;"></div><div class="sf sf1" style="top:-2.75em; left:41.90vw; animation-delay:-17.69s;"></div><div class="sf sf0" style="top:-1.48em; left:4.74vw; animation-delay:-16.27s;"></div><div class="sf sf0" style="top:-1.08em; left:31.11vw; animation-delay:-6.88s;"></div><div class="sf sf2" style="top:-3.10em; left:43.42vw; animation-delay:-4.78s;"></div><div class="sf sf1" style="top:-2.67em; left:91.86vw; animation-delay:-16.98s;"></div><div class="sf sf1" style="top:-2.34em; left:11.41vw; animation-delay:-0.99s;"></div><div class="sf sf1" style="top:-2.21em; left:19.71vw; animation-delay:-16.08s;"></div><div class="sf sf3" style="top:-4.76em; left:18.84vw; animation-delay:-15.10s;"></div><div class="sf sf2" style="top:-3.59em; left:12.10vw; animation-delay:-5.16s;"></div><div class="sf sf0" style="top:-1.44em; left:91.96vw; animation-delay:-1.13s;"></div><div class="sf sf0" style="top:-1.63em; left:85.30vw; animation-delay:-7.05s;"></div><div class="sf sf0" style="top:-1.41em; left:7.41vw; animation-delay:-3.04s;"></div><div class="sf sf1" style="top:-2.61em; left:77.48vw; animation-delay:-9.50s;"></div><div class="sf sf3" style="top:-4.53em; left:20.22vw; animation-delay:-12.37s;"></div><div class="sf sf0" style="top:-1.94em; left:70.25vw; animation-delay:-5.27s;"></div><div class="sf sf2" style="top:-3.39em; left:20.25vw; animation-delay:-11.82s;"></div><div class="sf sf2" style="top:-3.30em; left:19.75vw; animation-delay:-17.56s;"></div><div class="sf sf1" style="top:-2.26em; left:18.74vw; animation-delay:-12.65s;"></div><div class="sf sf2" style="top:-3.40em; left:67.80vw; animation-delay:-6.01s;"></div><div class="sf sf3" style="top:-4.81em; left:36.40vw; animation-delay:-18.15s;"></div><div class="sf sf1" style="top:-2.62em; left:92.22vw; animation-delay:-15.12s;"></div><div class="sf sf0" style="top:-1.64em; left:11.22vw; animation-delay:-0.11s;"></div><div class="sf sf3" style="top:-4.07em; left:49.26vw; animation-delay:-14.95s;"></div><div class="sf sf3" style="top:-4.30em; left:8.61vw; animation-delay:-1.02s;"></div><div class="sf sf3" style="top:-4.85em; left:64.56vw; animation-delay:-12.08s;"></div><div class="sf sf0" style="top:-1.92em; left:77.90vw; animation-delay:-10.47s;"></div><div class="sf sf0" style="top:-1.77em; left:67.53vw; animation-delay:-17.64s;"></div><div class="sf sf2" style="top:-3.60em; left:28.17vw; animation-delay:-11.23s;"></div><div class="sf sf1" style="top:-2.13em; left:92.19vw; animation-delay:-9.15s;"></div><div class="sf sf0" style="top:-1.60em; left:8.03vw; animation-delay:-7.79s;"></div><div class="sf sf3" style="top:-4.36em; left:5.67vw; animation-delay:-12.67s;"></div><div class="sf sf0" style="top:-1.06em; left:26.90vw; animation-delay:-18.02s;"></div><div class="sf sf1" style="top:-2.28em; left:18.39vw; animation-delay:-2.69s;"></div><div class="sf sf0" style="top:-1.33em; left:47.10vw; animation-delay:-7.38s;"></div><div class="sf sf2" style="top:-3.31em; left:47.92vw; animation-delay:-8.85s;"></div><div class="sf sf1" style="top:-2.58em; left:33.73vw; animation-delay:-14.20s;"></div><div class="sf sf2" style="top:-3.85em; left:55.80vw; animation-delay:-10.31s;"></div><div class="sf sf2" style="top:-3.77em; left:20.42vw; animation-delay:-14.65s;"></div><div class="sf sf0" style="top:-1.62em; left:34.45vw; animation-delay:-11.29s;"></div><div class="sf sf1" style="top:-2.54em; left:31.70vw; animation-delay:-6.03s;"></div><div class="sf sf1" style="top:-2.75em; left:23.20vw; animation-delay:-5.02s;"></div><div class="sf sf3" style="top:-4.53em; left:23.41vw; animation-delay:-8.86s;"></div><div class="sf sf1" style="top:-2.37em; left:64.92vw; animation-delay:-6.54s;"></div><div class="sf sf1" style="top:-2.61em; left:4.95vw; animation-delay:-19.17s;"></div><div class="sf sf1" style="top:-2.70em; left:36.66vw; animation-delay:-10.18s;"></div><div class="sf sf0" style="top:-1.01em; left:2.48vw; animation-delay:-5.43s;"></div><div class="sf sf1" style="top:-2.64em; left:84.31vw; animation-delay:-1.89s;"></div><div class="sf sf1" style="top:-2.81em; left:29.27vw; animation-delay:-17.14s;"></div><div class="sf sf3" style="top:-4.45em; left:68.35vw; animation-delay:-13.12s;"></div><div class="sf sf3" style="top:-4.62em; left:39.49vw; animation-delay:-15.81s;"></div><div class="sf sf2" style="top:-3.51em; left:8.87vw; animation-delay:-18.87s;"></div><div class="sf sf3" style="top:-4.91em; left:93.32vw; animation-delay:-13.49s;"></div><div class="sf sf0" style="top:-1.64em; left:84.79vw; animation-delay:-19.81s;"></div><div class="sf sf2" style="top:-3.67em; left:41.20vw; animation-delay:-8.53s;"></div><div class="sf sf3" style="top:-4.26em; left:73.08vw; animation-delay:-6.03s;"></div><div class="sf sf3" style="top:-4.12em; left:26.35vw; animation-delay:-4.04s;"></div><div class="sf sf2" style="top:-3.51em; left:30.08vw; animation-delay:-18.23s;"></div><div class="sf sf2" style="top:-3.01em; left:29.01vw; animation-delay:-0.50s;"></div><div class="sf sf3" style="top:-4.25em; left:57.77vw; animation-delay:-1.09s;"></div><div class="sf sf2" style="top:-3.80em; left:27.28vw; animation-delay:-10.61s;"></div><div class="sf sf0" style="top:-1.97em; left:83.62vw; animation-delay:-3.12s;"></div><div class="sf sf1" style="top:-2.14em; left:63.11vw; animation-delay:-12.20s;"></div><div class="sf sf2" style="top:-3.91em; left:73.21vw; animation-delay:-17.54s;"></div><div class="sf sf0" style="top:-1.46em; left:33.33vw; animation-delay:-12.42s;"></div><div class="sf sf0" style="top:-1.74em; left:30.32vw; animation-delay:-19.43s;"></div><div class="sf sf3" style="top:-4.37em; left:93.76vw; animation-delay:-5.68s;"></div><div class="sf sf0" style="top:-1.70em; left:84.38vw; animation-delay:-15.78s;"></div><div class="sf sf2" style="top:-3.33em; left:85.60vw; animation-delay:-6.58s;"></div><div class="sf sf0" style="top:-1.55em; left:25.17vw; animation-delay:-7.36s;"></div>
</main>
